design.md

```diff
@@ -0,0 +1,226 @@
+service JobworkerService {
+    rpc Start(Job) returns (JobID) {}
+    rpc Stop(JobID) returns (StopResponse) {}
+    rpc Status(JobID) returns (StatusResponse) {}
+    rpc StdOut(JobID) returns (stream OutputChunk) {}
+    rpc StdErr(JobID) returns (stream OutputChunk) {}
+}
+```
+
+### Authentication
+[mTLS](https://en.wikipedia.org/wiki/Mutual_authentication#mTLS) will be used, each user will have its own [X.509 certificate](https://en.wikipedia.org/wiki/X.509) ([PEM format](https://en.wikipedia.org/wiki/Privacy-Enhanced_Mail)), and the server too. Following the latest security standards, only [TLS v1.3](https://en.wikipedia.org/wiki/Transport_Layer_Security#TLS_1.3) with the following ciphersuites will be allowed:
+
+- `TLS_AES_128_GCM_SHA256`
+- `TLS_AES_256_GCM_SHA384`
+- `TLS_CHACHA20_POLY1305_SHA256`
+
+To simplify the testing process some pregenerated certificates will be provided, these are deemed unsafe for other purposes as they have been publicly exposed. New and safe certificates (RSA 2048-bit) can be generated with the following commands, using the [certstrap](https://github.com/square/certstrap) (or any similar) tool:
+
+#### Initialize a new certificate authority (to sign/verify server/client certificates)
+
+```
+$ certstrap init --common-name ca
+Enter passphrase (empty for no passphrase):
+Enter same passphrase again:
+Created out/ca.key
+Created out/ca.crt
+Created out/ca.crl
+```
+
+Only `ca.crt` is needed by the server or client to verify signatures. Do not distribute the other files.
+
+#### Request a new certificate for a client
+
+```
+$ certstrap request-cert --common-name user
+Enter passphrase (empty for no passphrase):
+Enter same passphrase again:
+Created out/user.key
+Created out/user.csr
+```
+
+`user.key` should be held only by the client and should be not distributed otherwise.
+
+#### Request a new certificate for a server
+
+```
+$ certstrap request-cert --domain localhost
+Enter passphrase (empty for no passphrase):
+Enter same passphrase again:
+Created out/localhost.key
+Created out/localhost.csr
+```
+
+`localhost.key` should be held only by the server and should not be distributed otherwise.
+
+#### Sign a requested certificate so it can be trusted by the server
+
+```
+$ certstrap sign user --CA ca
+Created out/user.crt from out/user.csr signed by out/ca.key
+```
+
+### Authorization
+
+The client will check the server provided certificate against his own known server certificate (provided by the sysadmin). The server will have a directory with the authorized client certificates. This is done so revoking access becomes possible without adding another service providing [CRL](https://en.wikipedia.org/wiki/Certificate_revocation_list) / [OCSP](https://en.wikipedia.org/wiki/Online_Certificate_Status_Protocol) or other mechanisms.
+
+To keep things simple, each job is considered to be owned by the user that started it, users are not able to interact in any way through the API with jobs not started by them.
```

Comment:
How will the server identify which client started the job and which client is calling the API?

Reply:
The server will use the user-provided certificate to know who the user is, different client invocations for the same user will still correspond to that user.

Perhaps this process-owner mechanism is not something desired, if so please let me know and I will make the jobs accesible for every authenticated user.